### ISSUE TITLE
Redesigning the color picker

### ISSUE THREAD

**Issue by dwelle** [excalidraw/excalidraw#5931]
Playground: https://codesandbox.io/s/cocky-cohen-f8lris?file=/src/App.js

![image](https://user-images.githubusercontent.com/5153846/204290680-848da27c-0081-4b3f-84f0-8a8df16256a0.png)

The basic principle:

![excal-color-picker-redesign-poc](https://user-images.githubusercontent.com/5153846/204362627-90a1af8c-054a-4af9-b186-d4827b4e7a07.gif)

### The idea

1. To address the issue of having too small choice of colors (to cover most use cases), we'll introduce a shade color picker. A row of ~5 shades for each base color.

2. Revamp the default colors. Either we just change shades, or we use a different color swatch altogether (we are currently using https://yeun.github.io/open-color/).

3. Use the same base colors for both the stroke and background.

    First off, both should support _white_, _black_, and _transparent_ colors. Also, you should be able to use the same color for both without copy/pasting the HEX value.

4. Probably still use a different shade for stroke and background.

    While we will use the same base colors, the default shade for background will be ~one shade lighter than the stroke. Stroke needs more contrast, while background needs to be less strong.
    
    Also, the common case is that you create a container with a background, and want to pick a matching stroke color which is still distinguishable from the background.

5. Show your most used colors even outside the color popup so you don't have to open it every time.

    Below, (1) is your "quick palette". Click to set color. (2) shows active color. If you want more options, you click on (2) to open the color popup.

    ![image](https://user-images.githubusercontent.com/5153846/204294668-70baec30-7756-4e18-b2b7-220866b739d1.png)

### FAQ

- **question:** Why a preset of shades instead of a full-blown color picker (wheel) allowing you to pick any color?
    - **answer:** Excalidraw's principle is simplicity and low friction. We want to give you a curated set of tools to create diagrams quickly without having to obsess about details.

        As such, we want to pick a set of colors that all look good by default, so you don't have to spend time on picking them yourselves.
        
        On top of that, a limited selection promotes consistency.

        That said, we still leave the option to add a custom color picker open.

- **question:** Why just 5 shades?
    - **answer:** While we could add more shades, such as 10 per color, it would again create too much choice (13 colors x 10 shades). If we pick our 5 shades well it should cover most use cases.

- **question:** What about keyboard shortcuts?
    - **answer:** Keyboard shortcuts will still be a thing! Sadly, we'll have to break them 🙁 (for those who have them in their muscle memory). Refer below:
    
        ![image](https://user-images.githubusercontent.com/5153846/204296254-1d9c4b9f-be35-4bb1-a20a-bb0064183003.png)

### Open questions

1. Will it be obvious (enough) how to open the color picker? The first five colors sets the active color, the 6th (active) color opens the color picker. How to make it more obvious?

2. Once you pick a shade, should the shade become the default for new colors as well (until you change it again)? I think it should.

3. What base colors to pick, and what shades? Especially the lightest and darkest shades.

5. Allow to customize "top picks"? Currently (on excalidraw.com), it uses the last 5 colors used on canvas that are not the default colors.

----

fix #4842
should take into account: #3643 #5872 #4086
WONTFIX for now: #5582 (full blown color pickers/wheel)

**Comment by tyfiero** [comment]
As a user, I would love this feature addition. The lack of colors bums me out. The shades idea is great, I wonder how dark mode would work with shades like that is my only concern.

**Comment by dwelle** [comment]
> I wonder how dark mode would work with shades like that is my only concern.

Good point. I've added dark mode into the playground: https://codesandbox.io/s/cocky-cohen-f8lris?file=/src/App.js

**Comment by dwelle** [comment]
I'm getting worried the color picker is getting too complex. New users might not know what to click on. 3 different sections of rows of colors!

Perhaps we don't have to show the "Top picks" and keep them only in the left sidebar?

![image](https://user-images.githubusercontent.com/5153846/204493574-0eb7586f-be37-4a71-8ffe-bc09eed2c8f0.png)

**Comment by ad1992** [comment]
> I'm getting worried the color picker is getting too complex. New users might not know what to click on. 3 different sections of rows of colors!
> 
> Perhaps we don't have to show the "Top picks" and keep them only in the left sidebar?
> 
> ![image](https://user-images.githubusercontent.com/5153846/204493574-0eb7586f-be37-4a71-8ffe-bc09eed2c8f0.png)

Yeah I agree lets not show the top pics on the popup to keep its simple

**Comment by anatoliibakhtin** [comment]
Hello ! 

You have a discussion here, I am happy to offer my options. I'll start from the outside-in order 1-9, answering your ideas (!), questions (?), considering (#) wishes. Probably most of the following information will be useless for you, I hope your time will not be wasted, but I was interested to try to organize it all on a sheet, using your tool and the excalidraw-obsidian plugin. Plan:

<img src="https://user-images.githubusercontent.com/71871141/205384006-c85048ea-9c3c-40ac-8ed6-46b8c70c8fd7.svg" width="100%">

# 1. #4086
(The keyboard shortcut prompts are not everywhere)
Add the "s,g" designations there — good idea.

<img src="https://user-images.githubusercontent.com/71871141/205371708-408fb612-4112-415a-8ff7-d1def5a249eb.svg" width="20%">

# 2. idea 3, 4 
(using different colors for canvas background, fills, strokes): 

<img src="https://user-images.githubusercontent.com/71871141/205371819-87462098-cce9-4a06-ae11-e7f0ad6ab599.svg" width="100%">

If you change only brightness (not saturation or hue), you get **1**, but if you also change saturation, you get **2** more contrasty, readable, less juicy. I would choose the 2nd option, but not knowing about 2 I would love to use 1. Picked up color saturation based on samples from [google cursive](https://cursive.apps.chrome/) 

<img src="(https://user-images.githubusercontent.com/71871141/205371878-58e8e457-6858-43e6-93c1-954513e4caf4.png" width="100%">

# 3. idea 5 
( move the section on colors used on the canvas to main menu) 
For me, this means the user's color is more important than the preset. It seems it might break the color style of the whole board. Drop by drop the mess will appear. If you do want to move the used colors section on the canvas, to the main menu, I like options 2, 3 (2 is a little harder to hit with your fingers than 3)

<img src="https://user-images.githubusercontent.com/71871141/205371927-07011473-b49f-40ea-bfba-2a937b5c74c2.svg" width="75%">

I would leave the layout unchanged, changing the color selection concept.

<img src="https://user-images.githubusercontent.com/71871141/205371984-206489e7-14d8-4007-9e18-22a7b2f3ac69.svg" width="20%">

 The concept —  pressing the S or G key pops up a color selection menu (as before or on the side), the cursor turns into a dropper, while pointing such a dropper at any button or tool, it is replaced by the cursor and back dropper if focus is lost.
# 4.  question 1 
(how to make the button on color selection even more obvious) 
Remove the frame around frequently used colors so that they do not look like a button (let the rule of your program will be everything in the frame — a button — can be pressed).
# 5. Should the colors used on the canvas be removed from the popup menu? 
I think so, all information that is duplicated should disappear.
# 6. idea 2, question 3 
(default color), (what are the base colors). 
Yes I would like other colors. In 6 months of using excalidraw I only used these colors (the ones with the green dot) as the others were too similar to each other, i.e. in fact the available to me were 6 groups— 10 colors.

<img src="https://user-images.githubusercontent.com/71871141/205372042-203238d2-daf0-45cf-bb32-82d58104e128.svg" width="100%">

  I was using google keep before, and after observing myself I made a conclusion, I used gray-black colors most often, + a few colors, to highlight, denote objects: important, ready, grass, water, fire., so I would like to have in excalidraw such a scheme: 

<img src="https://user-images.githubusercontent.com/71871141/205372149-89633b2b-fde4-4f05-ad57-9e93d9faaef2.svg" width="100%">

I took it from google cursive, reworked it a bit, black 3, 1 white, 1 transparent (#4842), colored 7x3 (almost like in a rainbow, so when drawing you can identify the object on the canvas with an object from real life, separate, set up a hierarchy), total of 25 + 1 transparent. Now I think — no more is needed (13x5 is too much, will take up space, I think so).
# 7. idea 1 
(few colors to choose from, way to expand) 
Shadow option — cool idea! I would reduce them from 5 to 3, build them into new palette.
# 8. question 2 
(would you save shade position when you change color for another color)
If my palette is 7x3— no need save shade position, each hue is a separate felt-tip (press S, choose color); if your palette is 13x5— yes, you should save, because by choosing another hue you kind of choose another ink mode.
# 9. question 4 
(whether to keep the ability to change the color) 
Yes, the user should be able to choose the color. Sometimes this is what the customer or my inner artist wants :).
    The palette itself will look like this: 

<img src="https://user-images.githubusercontent.com/71871141/205372420-ec8c899d-97c7-4630-9edf-a636a0045e5b.svg" width="50%">

 at the very top is the name of the mode in which the palette works, more details below. The components of the palette are arranged in 16 cells: 
- 1-J—  can change, depending on the palette mode (there is an internal rule №1—  

<img src="https://user-images.githubusercontent.com/71871141/205372665-eb6f7186-6d7c-45a9-b683-8bdb30674eac.png" width="50%"> 

**1** cells are always the darkest, most saturated, **J** the lightest, faded); 

- Z-B—  never change; 

- N— dropper (#5582): 

<img src="https://user-images.githubusercontent.com/71871141/205372756-705ba18a-5b33-43db-bf11-6a4e539b8bc1.svg" width="75%"> 

select objects, press N or on button, cursor changes to dropper, point it to pop up color palette, or object on canvas, or behind canvas (in other program), click (if you dragged cursor from dropper release mouse), objects will paint completely (stroke, fill) in 1 color, according to idea 3,4 (point 2 of current messag). There are only 2 droppers, 1 for color selection, the other for complete repainting of the object; 
- M— color palette settings: 

<img src="https://user-images.githubusercontent.com/71871141/205372797-dd3b06b2-84da-4738-992d-583806c7ab48.svg" width="35%">

1.  M1— its mode of operation (click to activate, inactive options become grayed out so as not to attract attention). Palette modes: 

<img src="https://user-images.githubusercontent.com/71871141/205372850-160428ed-a6f7-4b14-8986-afd3e097f113.svg" width="100%">

- **St**andard— you can change 1-7 cells, Q-J (see internal rule №1) — are auto-selectable tones for the hue above;
- **A**nalogous—Initially the same colors as Standart, but only cell 1 can be changed, everything else will be picked up automatically according to internal Rule №1, except for line 1-7 changes to the right according to the next color tone, not the saturation (Hue) 
- **M**onochromatic— only the center element "r" can be changed, the rest will be picked up according to rule №1;
- **C**omplementary, **S**plit **C**omplementary, **D**ouble **S**plit Complementary, **T**riad, **Sq**uare: you can change any single cell, the rest will be selected on the base of their scheme and rule №1; 
- **Cu**stom— all rules are disabled, cells are just containers for color. The frame is intentionally made bright black for demonstration only, in the working variant it will be as point 9 of current messag, barely visible (image above) 

2. M2— dumps all colors to default settings, sets the palette to Standard;
3. M3— import palette; 
4. M4— export the current palette (#3643). 

To summarize, the palette looks like this: 

<img src="https://user-images.githubusercontent.com/71871141/205372901-e3a5edb1-73f2-4d51-9f33-d6b5cfc81bc1.svg" width="75%">

 The old mode of arrangement of elements in the main menu; added s/g hint; new concept of color selection; different modes of palette operation; new color scheme; dropper tool for full recoloring, considering the principle "background, fill, outline"; palette settings: mode, reset, import, export; used colors on the canvas.

<img src="https://user-images.githubusercontent.com/71871141/205372934-21c3f0ed-19e0-4617-9955-90222d24e4ae.svg" width="75%">

 Horizontal or vertical. #5872 🤷‍♂️
 
 p/s Take this message as options, but not as a requirement. Thank you for your attention ! 🌼

**Comment by andyfleming** [comment]
The new color picker is looking great. I like the shades feature.

It would still be great to be able to select a color with a full color picker (or to be able to sample one from on screen/canvas). Clicking the swatch by the color text input seems like a reasonable place to add that option. I don't think it takes away from the simplicity of the defaults.

<img width="512" alt="Screen Shot 2023-03-25 at 4 55 26 PM" src="https://user-images.githubusercontent.com/721038/227747939-2f7eeaa5-fbc8-46fe-807c-51b758d8ed0c.png">

**Comment by andyfleming** [comment]
> @andyfleming we will definitely be adding an "eye dropper" to support sampling colors from the canvas.

Oh, perfect. That, combined with the color shade options would ease things a lot for me.

> But we want to avoid a full color picker for now to retain simplicity, reduce decision paralysis, and promote consistency!

Understandable. I think the swatches and shade features do that. Why even let someone type in a hex color then? (PLEASE DON'T TAKE THAT AWAY 😄 )

I think at the point that they are setting a custom color, they are already going out of their way to do something custom. Having a flexible picker available doesn't compromise that. I'm not suggesting you make it the first thing that comes up; just that it's available _somewhere_. Plus it would give you the eye dropper feature for "free".

**Comment by karlhorky** [comment]
Any PR or ongoing work that we can follow here?

Just adding an `<input type="color" />` color picker would already be a huge improvement without coming out with the polished final version.

Already "just works" in the browser if you add `type="color"` to the existing input:

![Screenshot 2023-04-18 at 10 48 37](https://user-images.githubusercontent.com/1935696/232725212-39cef502-60f6-4e03-ab06-d630304eba8d.png)

**Comment by anatoliibakhtin** [comment]
https://github.com/excalidraw/excalidraw/issues/6600#issuecomment-1554550519
> With the redesign we wanted you to have 5 quick-pick colors right at your fingertips without needing you to open the color picker itself.

We have a custom panel in obsidian-excalidraw with drawing pens up to 10 (straight ahead, without even opening the panel on the left) and soon there will be a [function to repaint](https://github.com/zsviczian/obsidian-excalidraw-plugin/issues/1125) (change color properties, size.,) objects by selecting one or the other pen. 

![image](https://github.com/excalidraw/excalidraw/assets/71871141/700dc16b-769d-4312-ba12-339022bb0366)
(just sharing good news)

**Comment by dwelle** [comment]
Closing as the redesign has been shipped, fulfilling its main purpose. We may make tweaks and incorporate some of the suggestions here in the future.
